File: app/schemas/lead.py
import uuid
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, Field

from app.models.booking_drop_off import LeadStatus, CustomerType


class ServiceInfo(BaseModel):
    id: uuid.UUID
    name: str
    model_config = {"from_attributes": True}


class AdminUserInfo(BaseModel):
    id: uuid.UUID
    name: str
    model_config = {"from_attributes": True}


class CustomerInfo(BaseModel):
    id: uuid.UUID
    name: Optional[str] = None
    phone: str
    model_config = {"from_attributes": True}


class LeadRead(BaseModel):
    id: uuid.UUID
    phone: str
    customer_id: Optional[uuid.UUID] = None
    customer: Optional[CustomerInfo] = None
    dropped_at_step: str
    selected_service_id: Optional[uuid.UUID] = None
    service: Optional[ServiceInfo] = None
    selected_slot_id: Optional[uuid.UUID] = None
    session_started_at: Optional[datetime] = None
    dropped_at: datetime
    status: LeadStatus
    customer_type: CustomerType
    assigned_to_id: Optional[uuid.UUID] = None
    assigned_to: Optional[AdminUserInfo] = None
    crm_notes: Optional[str] = None
    converted_appointment_id: Optional[uuid.UUID] = None
    
    # SLA tracking
    first_contacted_at: Optional[datetime] = None
    last_contacted_at: Optional[datetime] = None
    follow_up_at: Optional[datetime] = None
    
    # Lead scoring
    priority_score: Optional[int] = None
    
    created_at: datetime
    model_config = {"from_attributes": True}


class LeadUpdate(BaseModel):
    status: Optional[LeadStatus] = None
    assigned_to_id: Optional[uuid.UUID] = None
    crm_notes: Optional[str] = None
    follow_up_at: Optional[datetime] = None
    priority_score: Optional[int] = None


class LeadBulkUpdateRequest(BaseModel):
    lead_ids: list[uuid.UUID]
    status: Optional[LeadStatus] = None
    assigned_to_id: Optional[uuid.UUID] = None
    crm_notes: Optional[str] = None
    follow_up_at: Optional[datetime] = None


class LeadBulkAssignRequest(BaseModel):
    lead_ids: list[uuid.UUID]
    assigned_to_id: uuid.UUID


class LeadListResponse(BaseModel):
    items: list[LeadRead]
    total: int
    page: int
    page_size: int


class LeadConvertRequest(BaseModel):
    slot_id: uuid.UUID
    service_id: Optional[uuid.UUID] = None  # defaults to lead's service
    provider_id: Optional[uuid.UUID] = None
    notes: Optional[str] = None


# Activity schemas
class LeadActivityRead(BaseModel):
    id: uuid.UUID
    lead_id: uuid.UUID
    activity_type: str
    previous_value: Optional[str] = None
    new_value: Optional[str] = None
    notes: Optional[str] = None
    performed_by_id: Optional[uuid.UUID] = None
    performed_by: Optional[AdminUserInfo] = None
    performed_at: datetime
    metadata_json: Optional[str] = None
    model_config = {"from_attributes": True}


class LeadActivityCreate(BaseModel):
    lead_id: uuid.UUID
    activity_type: str
    previous_value: Optional[str] = None
    new_value: Optional[str] = None
    notes: Optional[str] = None
    performed_by_id: Optional[uuid.UUID] = None
    metadata_json: Optional[str] = None


class LeadActivityListResponse(BaseModel):
    items: list[LeadActivityRead]
    total: int

File: crm-frontend/src/types/lead.ts
export type LeadStatus = 'new_lead' | 'contacted' | 'follow_up' | 'converted' | 'lost';
export type CustomerType = 'prospect' | 'returning' | 're_engaged';
export type LeadActivityType = 
  | 'status_changed'
  | 'call_logged'
  | 'assigned'
  | 'unassigned'
  | 'converted'
  | 'note_added'
  | 'created'
  | 'follow_up_scheduled'
  | 'sla_breached'
  | 'reassigned';

export interface LeadServiceInfo {
  id: string;
  name: string;
}

export interface LeadCustomerInfo {
  id: string;
  name: string | null;
  phone: string;
}

export interface LeadAssignedTo {
  id: string;
  name: string;
}

export interface Lead {
  id: string;
  phone: string;
  customer_id: string | null;
  customer: LeadCustomerInfo | null;
  dropped_at_step: string;
  selected_service_id: string | null;
  service: LeadServiceInfo | null;
  selected_slot_id: string | null;
  session_started_at: string | null;
  dropped_at: string;
  status: LeadStatus;
  customer_type: CustomerType;
  assigned_to_id: string | null;
  assigned_to: LeadAssignedTo | null;
  crm_notes: string | null;
  converted_appointment_id: string | null;
  
  // SLA tracking
  first_contacted_at: string | null;
  last_contacted_at: string | null;
  follow_up_at: string | null;
  
  // Lead scoring
  priority_score: number | null;
  
  created_at: string;
}

export interface LeadListResponse {
  items: Lead[];
  total: number;
  page: number;
  page_size: number;
}

export interface LeadActivity {
  id: string;
  lead_id: string;
  activity_type: LeadActivityType;
  previous_value: string | null;
  new_value: string | null;
  notes: string | null;
  performed_by_id: string | null;
  performed_by: LeadAssignedTo | null;
  performed_at: string;
  metadata_json: string | null;
}

export interface LeadActivityListResponse {
  items: LeadActivity[];
  total: number;
  page: number;
  page_size: number;
}

export interface ActivityEvent {
  type: 'appointment_event' | 'profile_change' | 'message_sent';
  event: string;
  detail: string | null;
  source: string | null;
  changed_by_name: string | null;
  changed_by_email: string | null;
  appointment_id: string | null;
  created_at: string;
}
